### ISSUE TITLE
Contradictory error message

### ISSUE THREAD

**Issue by Zearin** [webdriverio/webdriverio#4541]
**Environment**
 - **WebdriverIO version:** [e.g. 5.13.2]
 - **Mode:** WDIO Testrunner
 - **If WDIO Testrunner, running sync/async:** sync
 - **Node.js version:** 12.10.0
 - **NPM version:** 6.11.3
 - **Browser name and version:** Version 77.0.3865.90 (Official Build) (64-bit)
 - **Platform name and version:** macOS Mojave (10.14.6)
 - **Additional wdio packages used (if applicable):** 
    - `wdio-chromedriver-service@5.0.2`
    - `chromedriver@77.0.0`
    - `@wdio/static-server-service@5.13.2`

**Config of WebdriverIO**

```javascript

const path = require('path')

const EXTENSION_PATH = path.resolve(__dirname, './build')

const IS_DEBUG_MODE = process.env['DEBUG'] !== undefined

//if (IS_DEBUG_MODE) console.log('\nRunning in debug mode.\n')

/*
 * ====================
 * Runner Configuration
 * ====================
 *
 * WebdriverIO allows it to run your tests in arbitrary locations (e.g. locally or
 * on a remote machine).
 */
const RUNNER_CONFIG = {
    debug: IS_DEBUG_MODE,
    
    runner: 'local',
    
    port: 9515, // default for chromedriver
    
    //
    // Override default path ('/wd/hub') for chromedriver service.
    path: '/',
}


/*
 * ==================
 * Specify Test Files
 * ==================
 * Define which test specs should run. The pattern is relative to the directory
 * from which `wdio` was called. Notice that, if you are calling `wdio` from an
 * NPM script (see https://docs.npmjs.com/cli/run-script) then the current working
 * directory is where your package.json resides, so `wdio` will be called from there.
 */
const TEST_FILES_CONFIG = {
    specs: [
        './test/specs/**/*.js'
    ],

    // Patterns to exclude.
    exclude: [],    
}


/*
 * ============
 * Capabilities
 * ============
 * Define your capabilities here. WebdriverIO can run multiple capabilities at the same
 * time. Depending on the number of capabilities, WebdriverIO launches several test
 * sessions. Within your capabilities you can overwrite the spec and exclude options in
 * order to group specific specs to a specific capability.
 *
 * First, you can define how many instances should be started at the same time. Let's
 * say you have 3 different capabilities (Chrome, Firefox, and Safari) and you have
 * set maxInstances to 1; wdio will spawn 3 processes. Therefore, if you have 10 spec
 * files and you set maxInstances to 10, all spec files will get tested at the same time
 * and 30 processes will get spawned. The property handles how many capabilities
 * from the same test should run tests.
 */
const CAPABILITIES_CONFIG = {
    maxInstances: 1,
    //
    // If you have trouble getting all important capabilities together, check out the
    // Sauce Labs platform configurator - a great tool to configure your capabilities:
    // https://docs.saucelabs.com/reference/platforms-configurator
    //
    capabilities: [{
        // maxInstances can get overwritten per capability. So if you have an in-house Selenium
        // grid with only 5 firefox instances available you can make sure that not more than
        // 5 instances get started at a time.
        maxInstances: 1,
        //
        browserName: 'chrome',
        
        // 'goog:chromeOptions': {
        //     'args': [`--load-extension=${EXTENSION_PATH}`]
        // }
        
        // If outputDir is provided WebdriverIO can capture driver session logs
        // it is possible to configure which logTypes to include/exclude.
        //
        // excludeDriverLogs: ['*'], // pass '*' to exclude all driver session logs
        // excludeDriverLogs: ['bugreport', 'server'],
    }],
}


/*
 * ===================
 * Test Configurations
 * ===================
 * Define all options that are relevant for the WebdriverIO instance here
 *
 * Level of logging verbosity: trace | debug | info | warn | error | silent
 */
const TEST_CONFIG = {
    logLevel: 'debug',
    
    /*
     * Set specific log levels per logger.
     *
     * Loggers are:
     * - webdriver, webdriverio
     * - @wdio/applitools-service, @wdio/browserstack-service, @wdio/devtools-service, @wdio/sauce-service
     * - @wdio/mocha-framework, @wdio/jasmine-framework
     * - @wdio/local-runner, @wdio/lambda-runner
     * - @wdio/sumologic-reporter
     * - @wdio/cli, @wdio/config, @wdio/sync, @wdio/utils
     * Level of logging verbosity: trace | debug | info | warn | error | silent
     */
    logLevels: {
        webdriver:                      'warn',
        '@wdio/static-server-service':  'debug',
        '@wdio/config':                 'debug'
        //,
        //'@wdio/applitools-service': 'info'
    },
    //
    // If you only want to run your tests until a specific amount of tests have failed use
    // bail (default is 0 - don't bail, run all tests).
    bail: 0,
    //
    // Set a base URL in order to shorten url command calls. If your `url` parameter starts
    // with `/`, the base url gets prepended, not including the path portion of your baseUrl.
    //
    // If your `url` parameter starts without a scheme or `/` (like `some/path`), the base url
    // gets prepended directly.
    //
    baseUrl: 'http://localhost:4567',
    //
    // Default timeout for all waitFor* commands.
    waitforTimeout: IS_DEBUG_MODE ? 10 * 60e3 : 30e3,
    
    //
    // Default timeout in milliseconds for request
    // if Selenium Grid doesn't send response
    //
    connectionRetryTimeout: IS_DEBUG_MODE ? 10 * 60e3 : 30e3,
    
    //
    // Default request retries count
    //
    connectionRetryCount: 1,
    
    //
    // Test runner services
    //
    // Services take over a specific job you don't want to take care of. They enhance
    // your test setup with almost no effort. Unlike plugins, they don't add new
    // commands. Instead, they hook themselves up into the test process.
    //
    services: ['chromedriver', 'static-server'],
    
    // Framework you want to run your specs with.
    // The following are supported: Mocha, Jasmine, and Cucumber
    // See also: https://webdriver.io/docs/frameworks.html
    //
    // Make sure you have the wdio adapter package for the specific framework installed
    // before running any tests.
    framework: 'mocha',
    
    //
    // The number of times to retry the entire specfile when it fails as a whole
    // specFileRetries: 1,
    //
    // Test reporter for stdout.
    // The only one supported by default is 'dot'
    // see also: https://webdriver.io/docs/dot-reporter.html
    //
    reporters: ['spec'],
    
    //
    // Options to be passed to Mocha.
    // See the full list at http://mochajs.org/
    mochaOpts: {
        ui: 'bdd',
        timeout: IS_DEBUG_MODE ? 10 * 60e3 : 30e3
    },
}


/*
 * ===================
 * Static Server Service Configuration
 * ===================
 *
 * @see https://webdriver.io/docs/static-server-service.html
 *
 * 
 */
const STATIC_SERVER_SERVICE_CONFIG = {
    'staticServerFolders': [
        { mount: '/', path: './test/data' }
    ],
    
    //  Default: 4567
    'staticServerPort': 4567,
    
    //  Debugging logs, will print mount points and requests. 
    //  When staticServerLogs `true`, it will print to the console. 
    //  Otherwise, a string is treated as the log folder.
    'staticServerLog': true,
    
    //  default: []
    //'staticServerMiddleware':
}

/*
 * =====
 * Hooks
 * =====
 * WebdriverIO provides several hooks you can use to interfere with the test process in order to enhance
 * it and to build services around it. You can either apply a single function or an array of
 * methods to it. If one of them returns with a promise, WebdriverIO will wait until that promise got
 * resolved to continue.
 */
const HOOKS_CONFIG = {
    /**
     * Gets executed once before all workers get launched.
     * @param {Object} config wdio configuration object
     * @param {Array.<Object>} capabilities list of capabilities details
     */
    // onPrepare: function (config, capabilities) {
    // },
    
    /**
     * Gets executed just before initialising the webdriver session and test framework. It allows you
     * to manipulate configurations depending on the capability or spec.
     * @param {Object} config wdio configuration object
     * @param {Array.<Object>} capabilities list of capabilities details
     * @param {Array.<String>} specs List of spec file paths that are to be run
     */
    // beforeSession: function (config, capabilities, specs) {
    // },
    
    /**
     * Gets executed before test execution begins. At this point you can access to all global
     * variables like `browser`. It is the perfect place to define custom commands.
     * @param {Array.<Object>} capabilities list of capabilities details
     * @param {Array.<String>} specs List of spec file paths that are to be run
     */
    before: function (capabilities, specs) {
        browser.setTimeout({
            'pageLoad': IS_DEBUG_MODE ? ( 1 * 60e3) : 30e3,
            'script':   IS_DEBUG_MODE ? (10 * 60e3) : 30e3
        })
    },
    
    /**
     * Runs before a WebdriverIO command gets executed.
     * @param {String} commandName hook command name
     * @param {Array} args arguments that command would receive
     */
    // beforeCommand: function (commandName, args) {
    // },
    
    /**
     * Hook that gets executed before the suite starts
     * @param {Object} suite suite details
     */
    // beforeSuite: function (suite) {
    // },
    
    /**
     * Function to be executed before a test (in Mocha/Jasmine) or a step (in Cucumber) starts.
     * @param {Object} test test details
     */
    // beforeTest: function (test) {
    // },
    
    /**
     * Hook that gets executed _before_ a hook within the suite starts (e.g. runs before calling
     * beforeEach in Mocha)
     */
    // beforeHook: function () {
    // },
    
    /**
     * Hook that gets executed _after_ a hook within the suite starts (e.g. runs after calling
     * afterEach in Mocha)
     */
    // afterHook: function () {
    // }, 
    
    /**
     * Function to be executed after a test (in Mocha/Jasmine) or a step (in Cucumber) starts.
     * @param {Object} test test details
     */
    // afterTest: function(test) {
    // },

    /**
     * Hook that gets executed after the suite has ended
     * @param {Object} suite suite details
     */
    // afterSuite: function (suite) {
    // },
    /**
     * Runs after a WebdriverIO command gets executed
     * @param {String} commandName hook command name
     * @param {Array} args arguments that command would receive
     * @param {Number} result 0 - command success, 1 - command error
     * @param {Object} error error object if any
     */
    // afterCommand: function (commandName, args, result, error) {
    // },
    
    /**
     * Gets executed after all tests are done. You still have access to all global variables from
     * the test.
     * @param {Number} result 0 - test pass, 1 - test fail
     * @param {Array.<Object>} capabilities list of capabilities details
     * @param {Array.<String>} specs List of spec file paths that ran
     */
    // after: function (result, capabilities, specs) {
    // },
    
    /**
     * Gets executed right after terminating the webdriver session.
     * @param {Object} config wdio configuration object
     * @param {Array.<Object>} capabilities list of capabilities details
     * @param {Array.<String>} specs List of spec file paths that ran
     */
    // afterSession: function (config, capabilities, specs) {
    // },
    
    /**
     * Gets executed after all workers got shut down and the process is about to exit. An error
     * thrown in the onComplete hook will result in the test run failing.
     * @param {Object} exitCode 0 - success, 1 - fail
     * @param {Object} config wdio configuration object
     * @param {Array.<Object>} capabilities list of capabilities details
     * @param {<Object>} results object containing test results
     */
    // onComplete: function(exitCode, config, capabilities, results) {
    // },
    
    /**
     * Gets executed when a refresh happens.
     * @param {String} oldSessionId session ID of the old session
     * @param {String} newSessionId session ID of the new session
     */
    //onReload: function(oldSessionId, newSessionId) {
    //}
}


exports.config = {
    ...RUNNER_CONFIG,
    ...TEST_FILES_CONFIG,
    ...CAPABILITIES_CONFIG,
    ...TEST_CONFIG,
    ...STATIC_SERVER_SERVICE_CONFIG,
    ...HOOKS_CONFIG
}

```

**Describe the bug**
I’m still learning WebdriverIO, so I fully acknowledge I am probably doing _something_ wrong. However, the error I’m getting contradicts itself, offering no direction for me to fix whatever is wrong.  

Here it is:

```
[chrome  mac os x #0-0] Error: Malformed type for "args" parameter of command executeScript
[chrome  mac os x #0-0] Expected: (string|object|number|boolean|undefined)[]
[chrome  mac os x #0-0] Actual: object
[chrome  mac os x #0-0] For more info see https://w3c.github.io/webdriver/#dfn-execute-script
[chrome  mac os x #0-0]     at processTicksAndRejections (internal/process/task_queues.js:93:5)
[chrome  mac os x #0-0]     at Context.<anonymous> (/Users/…/…/…/test/specs/test.js:77:36)
```

**To Reproduce**

[Include code or an example repository that can easily be set up]

**Expected behavior**
When I do something wrong, I expect the error message to (1) avoid contradicting itself; and (2) point me in the right direction to fix the error.  

**Log**
> If applicable, add logs to help explain your problem.

I am probably doing something wrong, but this bug report is about the error message itself.

**Additional context**
The bug I’m reporting here is the error message that says I need to pass an argument of type `(string|object|number|boolean|undefined)[]`, then _immediately_ says the error was caused by passing `object`.

What I am passing is a function.  I have tried declaring it as an arrow function, a `function someName() {…}`, and assigning the function to a variable `const someFunc = function() {…}`.

**Comment by christian-bromann** [comment]
Can you share the code from `test/specs/test.js:77:36`? It might be because the function expects an array of these types, not just a single type

**Comment by Zearin** [comment]
(I _think_ this is right. I’ve been fiddling with the code all morning trying to figure out what’s going on, so I manually restored )

A snippet of the relevant bits:

```javascript
function createCalendarScraper() {
    let calScraper
    calScraper = new CalendarScraper()
    console.log('CalendarScraper created!')
    return calScraper
}

const result = browser.execute( (fn) => {
    
    switch (document.readyState) {
        case 'loading':
            document.addEventListener('DOMContentLoaded', () => {
                return fn()
            })
            break
        
        case 'interactive':
            // The document has finished loading. 
            // We can now access the DOM elements.
            // But sub-resources such as images, stylesheets, and frames are still loading.
            return fn()
            break
        
        case 'complete':
            return fn()
            break
    }

}, createCalendarScraper )
```

**Comment by mgrybyk** [comment]
What is CalendarScraper?

I'd avoid passing complex objects to execute function.

**Comment by christian-bromann** [comment]
Yes, you can't pass in a function as argument. I will leave this open to fix the error message.

**Comment by Zearin** [comment]
CalendarScraper is a class I’m testing.  

I have to work on a website with some pretty outdated calendar software, so I wrote a class to help with things like formatting, marking schedule conflicts, etc.

----

If I change the above code to the following:

```javascript
function createCalendarScraper() {
    let calScraper
    calScraper = new CalendarScraper()
    console.log('CalendarScraper created!')
    return calScraper
}

const result = browser.execute( (fns) => {
    let fn = fns[0]
    
    switch (document.readyState) {
        case 'loading':
            document.addEventListener('DOMContentLoaded', () => {
                return fn()
            })
            break
        
        case 'interactive':
            // The document has finished loading. 
            // We can now access the DOM elements.
            // But sub-resources such as images, stylesheets, and frames are still loading.
            return fn()
            break
        
        case 'complete':
            return fn()
            break
    }

}, [createCalendarScraper] )

console.log(result)
```

Then I get the following error instead:

```
[0-0] 2019-09-27T17:07:00.553Z DEBUG webdriver: request failed due to response error: javascript error
[0-0] 2019-09-27T17:07:00.553Z ERROR webdriver: Request failed due to javascript error: javascript error: fn is not a function
  (Session info: chrome=77.0.3865.90)
    at getErrorFromResponseBody (/Users/amrogers/Projects/UnityOfFairfaxTools/node_modules/webdriver/build/utils.js:175:10)
    at Request._callback (/Users/amrogers/Projects/UnityOfFairfaxTools/node_modules/webdriver/build/request.js:110:64)
    at Request.self.callback (/Users/amrogers/Projects/UnityOfFairfaxTools/node_modules/request/request.js:185:22)
    at Request.emit (events.js:209:13)
    at Request.EventEmitter.emit (domain.js:476:20)
    at Request.<anonymous> (/Users/amrogers/Projects/UnityOfFairfaxTools/node_modules/request/request.js:1161:10)
    at Request.emit (events.js:209:13)
    at Request.EventEmitter.emit (domain.js:476:20)
    at IncomingMessage.<anonymous> (/Users/amrogers/Projects/UnityOfFairfaxTools/node_modules/request/request.js:1083:12)
    at Object.onceWrapper (events.js:298:28)
[0-0] javascript error in "Feature: Calendar Pages Scenario: When instantiating a CalendarScraper Then it should not throw"
javascript error: fn is not a function
  (Session info: chrome=77.0.3865.90)
```

Which still confuses me, because it is definitely a function...

**Comment by Zearin** [comment]
Right above the error I just posted, the debug output _says_ it is a function!  

(Sorry for not prepending this to the previous comment’s error text.)

```
[0-0] 2019-09-27T17:07:00.541Z INFO webdriver: COMMAND executeScript("return ((fns) => {
                let fn = fns[0]
                
                switch (document.readyState) {
                    case 'loading':
                        document.addEventListener('DOMContentLoaded', () => {
                            return fn()
                        })
                        break
                    
                    case 'interactive':
                        // The document has finished loading. 
                        // We can now access the DOM elements.
                        // But sub-resources such as images, stylesheets, and frames are still loading.
                        return fn()
                        break
                    
                    case 'complete':
                        return fn()
                        break
                }
            
            }).apply(null, arguments)", <object>)
[0-0] 2019-09-27T17:07:00.541Z INFO webdriver: [POST] http://127.0.0.1:9515/session/bc7a8a7c73bb45aae03e19747c06aa20/execute/sync
[0-0] 2019-09-27T17:07:00.541Z INFO webdriver: DATA {
  script: 'return ((fns) => {\n' +
    '                let fn = fns[0]\n' +
    '                \n' +
    '                switch (document.readyState) {\n' +
    "                    case 'loading':\n" +
    "                        document.addEventListener('DOMContentLoaded', () => {\n" +
    '                            return fn()\n' +
    '                        })\n' +
    '                        break\n' +
    '                    \n' +
    "                    case 'interactive':\n" +
    '                        // The document has finished loading. \n' +
    '                        // We can now access the DOM elements.\n' +
    '                        // But sub-resources such as images, stylesheets, and frames are still loading.\n' +
    '                        return fn()\n' +
    '                        break\n' +
    '                    \n' +
    "                    case 'complete':\n" +
    '                        return fn()\n' +
    '                        break\n' +
    '                }\n' +
    '            \n' +
    '            }).apply(null, arguments)',
  args: [ [ [Function: createCalendarScraper] ] ]
}
[0-0] 2019-09-27T17:07:00.548Z DEBUG webdriver: request failed due to response error: javascript error
[0-0] 2019-09-27T17:07:00.548Z WARN webdriver: Request failed due to javascript error: fn is not a function
  (Session info: chrome=77.0.3865.90)
[0-0] 2019-09-27T17:07:00.548Z INFO webdriver: Retrying 1/1
[0-0] 2019-09-27T17:07:00.548Z INFO webdriver: [POST] http://127.0.0.1:9515/session/bc7a8a7c73bb45aae03e19747c06aa20/execute/sync
[0-0] 2019-09-27T17:07:00.549Z INFO webdriver: DATA {
  script: 'return ((fns) => {\n' +
    '                let fn = fns[0]\n' +
    '                \n' +
    '                switch (document.readyState) {\n' +
    "                    case 'loading':\n" +
    "                        document.addEventListener('DOMContentLoaded', () => {\n" +
    '                            return fn()\n' +
    '                        })\n' +
    '                        break\n' +
    '                    \n' +
    "                    case 'interactive':\n" +
    '                        // The document has finished loading. \n' +
    '                        // We can now access the DOM elements.\n' +
    '                        // But sub-resources such as images, stylesheets, and frames are still loading.\n' +
    '                        return fn()\n' +
    '                        break\n' +
    '                    \n' +
    "                    case 'complete':\n" +
    '                        return fn()\n' +
    '                        break\n' +
    '                }\n' +
    '            \n' +
    '            }).apply(null, arguments)',
  args: [ [ [Function: createCalendarScraper] ] ]
}
[0-0] 2019-09-27T17:07:00.553Z DEBUG webdriver: request failed due to response error: javascript error
[0-0] 2019-09-27T17:07:00.553Z ERROR webdriver: Request failed due to javascript error: javascript error: fn is not a function
  (Session info: chrome=77.0.3865.90)
    at getErrorFromResponseBody (/Users/amrogers/Projects/UnityOfFairfaxTools/node_modules/webdriver/build/utils.js:175:10)
    at Request._callback (/Users/amrogers/Projects/UnityOfFairfaxTools/node_modules/webdriver/build/request.js:110:64)
    at Request.self.callback (/Users/amrogers/Projects/UnityOfFairfaxTools/node_modules/request/request.js:185:22)
    at Request.emit (events.js:209:13)
    at Request.EventEmitter.emit (domain.js:476:20)
    at Request.<anonymous> (/Users/amrogers/Projects/UnityOfFairfaxTools/node_modules/request/request.js:1161:10)
    at Request.emit (events.js:209:13)
    at Request.EventEmitter.emit (domain.js:476:20)
    at IncomingMessage.<anonymous> (/Users/amrogers/Projects/UnityOfFairfaxTools/node_modules/request/request.js:1083:12)
    at Object.onceWrapper (events.js:298:28)
````

**Comment by christian-bromann** [comment]
@Zearin yes, because you can serialize functions

**Comment by Zearin** [comment]
> @Zearin yes, because you can serialize functions

Can you explain?  (“Serialize” into what?  What—by contrast—is not serializable?)

**Comment by Zearin** [comment]
> You can use executeAsync and call done callback instead of calling createCalendarScraper inside execute

Is it possible to use `async/await` with `executeAsync` instead of callbacks?

**Comment by christian-bromann** [comment]
> Can you explain?

You can't serialize a function, meaning that you can not pass in a function as a argument to the execute command

**Comment by christian-bromann** [comment]
error message got fixed in #4556

**Comment by Zearin** [comment]
Thanks, @christian-bromann :)
